src/Stream.ts

```diff
@@ -0,0 +1,108 @@
+import { Effect } from "./Effect";
+import * as Fx from "./Fx";
+import * as Task from "./Task";
+import * as Async from "./Async";
+import * as Fail from "./Fail";
+import * as Fork from "./Fork";
+
+export class Stream<A> extends Effect('Stream')<A, void> { }
+
+export namespace Stream { 
+  export type Event<T> = T extends Stream<infer A> ? A : never
+}
+
+export function event<const A>(a: A): Fx.Fx<Stream<A>, void> { 
+  return new Stream(a)
+}
+
+export function observe<E, R, E2>(fx: Fx.Fx<E, R>, f: (a: Stream.Event<E>) => Fx.Fx<E2, void>): Fx.Fx<Exclude<E, Stream<any>> | E2, R> { 
+  return fx.pipe(Fx.handle<typeof Stream, E2>(Stream, (a) => f(a as Stream.Event<E>)))
+} 
+
+export function collectAll<E, A>(fx: Fx.Fx<E, A>): Fx.Fx<Exclude<E, Stream<any>>, readonly Stream.Event<E>[]> {
+  return Fx.fx(function* () {
+    const events: Stream.Event<E>[] = []
+    yield* observe(fx, a => { 
+      events.push(a)
+      return Fx.unit
+    })
+    return events
+  })
+}
+
+export function map<E, A, B>(fx: Fx.Fx<E, A>, f: (a: Stream.Event<E>) => B): Fx.Fx<Exclude<E, Stream<any>> | Stream<B>, A> { 
+  return observe(fx, a => event(f(a)))
+}
+
+export function filter<E, A>(fx: Fx.Fx<E, A>, f: (a: Stream.Event<E>) => boolean): Fx.Fx<Exclude<E, Stream<any>> | Stream<Stream.Event<E>>, A> {
+  return observe(fx, a => f(a) ? event(a) : Fx.unit)
+}
+
+export function flatMap<E, X, B, E2>(fx: Fx.Fx<E, X>, f: (a: Stream.Event<E>) => Fx.Fx<E2, B>): Fx.Fx<Fork.Fork | Async.Async | E | E2 | Stream<B>, X> {
```

Comment:
What are your thoughts on whether `f` should be required to return something like `Fx.Fx<E2 | Stream<B>, ???>`?

That seems more like a typical flatMap (although we don't have to be bound to that).  It's really unclear what it should do with the `???`, though ... probably just have to discard them.

Reply:
Another idea is maybe to rename this to something like `forkMap` ...

Thoughts?

Reply:
Oh hmm, I think I actually have that typed wrong. At the moment it does just discard those "inner" output values.  I'll have to try and fix up the signature here to match the actual behavior a bit better.

I do like the idea of `forkMap` which _does_ utilize those return values though. It seems like a logical way to want to structure your programs mostly using Fx and utilizing Stream really just for the concurrency patterns it offers for multi-shot data structures and events